File: backend/aerodue/inference/rationale.py
import json
from pathlib import Path

import yaml

from aerodue.domain.claim import ClaimRecommendation
from aerodue.domain.coverage import UserCoverageProfile
from aerodue.domain.disruption import DisruptionEvent


def load_model_config() -> dict:
    path = Path(__file__).parent / "models.yaml"
    with path.open() as f:
        return yaml.safe_load(f)


def build_rationale_prompt(
    event: DisruptionEvent,
    profile: UserCoverageProfile,
    claims: list[ClaimRecommendation],
    excerpts: str = "",
) -> str:
    """Format prompt for on-device LLM; no network call here."""
    cfg = load_model_config()
    template = cfg["prompts"]["claim_rationale"]
    return template.format(
        disruption_json=event.model_dump_json(),
        coverage_json=profile.model_dump_json(),
        claims_json=json.dumps([c.model_dump() for c in claims]),
        excerpts=excerpts or "(no excerpts loaded — bundle regulation snippets in APK)",
    )


def explain_offline(
    event: DisruptionEvent,
    profile: UserCoverageProfile,
    claims: list[ClaimRecommendation],
) -> str:
    """
    Stub when LLM runtime is not wired. Android will call native inference later.
    """
    if not claims:
        return "No eligible compensation paths found for this disruption with your current coverage."
    lines = [f"• {c.title}: {c.summary}" for c in claims[:5]]
    return "Offline rule assessment complete.\n\n" + "\n".join(lines)

File: android/core/src/main/kotlin/com/aerodue/core/inference/Rationale.kt
package com.aerodue.core.inference

import com.aerodue.core.domain.ClaimRecommendation
import com.aerodue.core.domain.DisruptionEvent
import com.aerodue.core.domain.UserCoverageProfile

object Rationale {

    fun buildPrompt(
        event: DisruptionEvent,
        profile: UserCoverageProfile,
        claims: List<ClaimRecommendation>,
        excerpts: String = "",
    ): String = """
        Disruption: $event
        Coverage: $profile
        Rule hits: $claims
        Relevant excerpts:
        ${excerpts.ifBlank { "(bundle regulation snippets in assets/regulations/)" }}
    """.trimIndent()

    /** Rule-based summary when LiteRT-LM model is not loaded. */
    fun explainOffline(
        event: DisruptionEvent,
        profile: UserCoverageProfile,
        claims: List<ClaimRecommendation>,
    ): String {
        if (claims.isEmpty()) {
            return "No eligible compensation paths found for this disruption with your current coverage."
        }
        val lines = claims.take(5).joinToString("\n") { "• ${it.title}: ${it.summary}" }
        return "Offline rule assessment complete.\n\n$lines"
    }

    fun explainWithoutLlm(prompt: String): String =
        "Offline LLM not loaded (${ModelProfiles.defaultProfileId}). " +
            "Rule-based assessment is active.\n\nPrompt preview:\n${prompt.take(200)}…"
}
